### ISSUE TITLE
[Question]: 在spu上运行MobileNetV3隐私推理，SPU每次输出都不一样

### ISSUE THREAD

**Issue by zsan2023** [secretflow/spu#1218]
### Issue Type

Build/Install

### Modules Involved

MPC protocol

### Have you reproduced the bug with SPU HEAD?

Yes

### Have you searched existing issues?

Yes

### SPU Version

0.9.0

### OS Platform and Distribution

ubuntu 22.04

### Python Version

3.10.14

### Current Behavior?

我在运行MobileNetV3，模型与run_on_spu代码如下，每次输出的结果都不同并且与cpu运行结果不一样（cpu输出是固定的）
MobileNetV3：
```
from typing import Union, Tuple, Optional
import jax
import flax.linen as nn
import jax.numpy as jnp


def _make_divisible(v, divisor=8, min_value=None):
    if min_value is None:
        min_value = divisor
    new_v = max(min_value, int(v + divisor / 2) // divisor * divisor)
    if new_v < 0.9 * v:
        new_v += divisor
    return new_v


def correct_pad(input_size: Union[int, tuple], kernel_size: int) -> Tuple[Tuple[int, int], Tuple[int, int]]:
    if isinstance(input_size, int):
        input_size = (input_size, input_size)
    kernel_size = (kernel_size, kernel_size)
    adjust = (1 - input_size[0] % 2, 1 - input_size[1] % 2)
    correct = (kernel_size[0] // 2, kernel_size[1] // 2)
    return ((correct[0] - adjust[0], correct[0]),
            (correct[1] - adjust[1], correct[1]))


def hard_sigmoid(x):
    return jax.nn.hard_sigmoid(x)


def hard_swish(x):
    return x * hard_sigmoid(x)


class SEBlock(nn.Module):
    filters: int
    se_ratio: float = 0.25
    prefix: str = "se_"

    @nn.compact
    def __call__(self, inputs):
        x = jnp.mean(inputs, axis=(1, 2), keepdims=True)
        x = nn.Conv(
            features=_make_divisible(self.filters * self.se_ratio),
            kernel_size=(1, 1),
            name=self.prefix + "conv1"
        )(x)
        x = nn.relu(x)
        x = nn.Conv(
            features=self.filters,
            kernel_size=(1, 1),
            name=self.prefix + "conv2"
        )(x)
        x = hard_sigmoid(x)
        return inputs * x


class InvertedResBlock(nn.Module):
    input_c: int
    kernel_size: int
    exp_c: int
    out_c: int
    use_se: bool
    activation: str
    stride: int
    alpha: float = 1.0
    block_id: int = 0
    @nn.compact
    def __call__(self, x, training=True):  # 添加training参数
        residual = x
        input_c = _make_divisible(self.input_c * self.alpha)
        exp_c = _make_divisible(self.exp_c * self.alpha)
        out_c = _make_divisible(self.out_c * self.alpha)
        act = nn.relu if self.activation == "RE" else hard_swish
        prefix = f"expanded_conv_{self.block_id}/" if self.block_id else "expanded_conv/"
        if self.block_id != 0:
            x = nn.Conv(
                features=exp_c,
                kernel_size=(1, 1),
                use_bias=False,
                name=prefix + "expand_conv"
            )(x)
            x = nn.BatchNorm(
                epsilon=1e-3,
                momentum=0.999,
                use_running_average=not training,  # 传入训练模式
                name=prefix + "expand_bn"
            )(x)
            x = act(x)

        # Depthwise convolution with padding
        if self.stride == 2:
            input_size = x.shape[1:3]
            padding = correct_pad(input_size, self.kernel_size)
            x = jnp.pad(x, ((0, 0), padding[0], padding[1], (0, 0)), mode="constant")
            conv_pad = "VALID"
        else:
            conv_pad = "SAME"

        x = nn.Conv(
            features=exp_c,
            kernel_size=(self.kernel_size, self.kernel_size),
            strides=(self.stride, self.stride),
            feature_group_count=exp_c,
            padding=conv_pad,
            use_bias=False,
            name=prefix + "depthwise_conv"
        )(x)
        x = nn.BatchNorm(
            epsilon=1e-3,
            momentum=0.999,
            use_running_average=not training,  # 添加参数
            name=prefix + "depthwise_bn"
        )(x)
        x = act(x)

        # Squeeze-excite
        if self.use_se:
            x = SEBlock(filters=exp_c, prefix=prefix + "se_")(x)

        # Projection
        x = nn.Conv(
            features=out_c,
            kernel_size=(1, 1),
            use_bias=False,
            name=prefix + "project_conv"
        )(x)
        x = nn.BatchNorm(
            epsilon=1e-3,
            momentum=0.999,
            use_running_average=not training, 
            name=prefix + "project_bn"
        )(x)

        # Residual connection
        if self.stride == 1 and input_c == out_c:
            x = x + residual
        return x


class MobileNetV3Small(nn.Module):
    num_classes: int = 1000
    alpha: float = 1.0
    include_top: bool = True
    input_shape: Tuple[int, int, int] = (224, 224, 3)

    @nn.compact
    def __call__(self, x, training=True):
        # Initial layer
        x = nn.Conv(
            features=16,
            kernel_size=(3, 3),
            strides=(2, 2),
            padding="SAME",
            use_bias=False,
            name="init_conv"
        )(x)
        x = nn.BatchNorm(
            epsilon=1e-3,
            momentum=0.999,
            use_running_average=not training,
            name="init_bn"
        )(x)
        x = hard_swish(x)

        # Inverted residual blocks
        x = InvertedResBlock(16, 3, 16, 16, True, "RE", 2, self.alpha, 0)(x, training=training)
        x = InvertedResBlock(16, 3, 72, 24, False, "RE", 2, self.alpha, 1)(x, training=training)
        x = InvertedResBlock(24, 3, 88, 24, False, "RE", 1, self.alpha, 2)(x, training=training)
        x = InvertedResBlock(24, 5, 96, 40, True, "HS", 2, self.alpha, 3)(x, training=training)
        x = InvertedResBlock(40, 5, 240, 40, True, "HS", 1, self.alpha, 4)(x, training=training)
        x = InvertedResBlock(40, 5, 240, 40, True, "HS", 1, self.alpha, 5)(x, training=training)
        x = InvertedResBlock(40, 5, 120, 48, True, "HS", 1, self.alpha, 6)(x, training=training)
        x = InvertedResBlock(48, 5, 144, 48, True, "HS", 1, self.alpha, 7)(x, training=training)
        x = InvertedResBlock(48, 5, 288, 96, True, "HS", 2, self.alpha, 8)(x, training=training)
        x = InvertedResBlock(96, 5, 576, 96, True, "HS", 1, self.alpha, 9)(x, training=training)
        x = InvertedResBlock(96, 5, 576, 96, True, "HS", 1, self.alpha, 10)(x, training=training)

        # Final layers
        last_c = _make_divisible(576 * self.alpha)
        x = nn.Conv(
            features=last_c,
            kernel_size=(1, 1),
            padding="SAME",
            use_bias=False,
            name="final_conv"
        )(x)
        x = nn.BatchNorm(
            epsilon=1e-3,
            momentum=0.999,
            use_running_average=not training,
            name="final_bn"
        )(x)
        x = hard_swish(x)

        if self.include_top:
            x = jnp.mean(x, axis=(1, 2))
            x = nn.Dense(
                features=_make_divisible(1024 * self.alpha),
                name="final_dense1"
            )(x)
            x = hard_swish(x)
            x = nn.Dense(
                features=self.num_classes,
                name="final_dense2"
            )(x)
            x = nn.softmax(x)
        return x
```

run_on_spu：
```
def run_on_spu(model, params, image):
    print(f"Running on SPU ...")

    def eval_fn(params, image):
        with hijack(enabled=True):
            logits = model.apply(params, image, training=False)
        return logits

    # Upload data and model to different parties
    spu_image = ppd.device("P1")(lambda x: x)(image)      # Input from P1
    spu_params = ppd.device("P2")(lambda x: x)(params)     # Params from P2

    start = time.time()
    logits_spu = ppd.device("SPU")(eval_fn, copts=copts)(spu_params, spu_image)
    end = time.time()

    logits = ppd.get(logits_spu)
    pred = jnp.argmax(logits, -1)

    print(f"SPU runtime: {(end - start):.3f}s")
    print(f"SPU prediction: {pred}, logits shape: {logits.shape}")

    return logits
```

### Standalone code to reproduce the issue

```Python
print("A bug")
```

### Relevant log output

```Shell

```

**Comment by deadlywing** [comment]
From the log output, it looks like you’re using the `cheetah` protocol. Some protocols can have random 1–2 bit errors, and as the computation depth increases, these errors are amplified step by step, so the final results may differ somewhat.

**Comment by zsan2023** [comment]
> From the log output, it looks like you’re using the `cheetah` protocol. Some protocols can have random 1–2 bit errors, and as the computation depth increases, these errors are amplified step by step, so the final results may differ somewhat.

我在运行resnet50时，是没有此类问题的。resnet50的计算量应该还要更大一些

**Comment by deadlywing** [comment]
Even a simple example of matrix multiplication can have some error after two runs.

```python
import jax.numpy as jnp
import numpy as np

np.random.seed(42)

import spu.libspu as libspu
import spu.utils.simulation as ppsim

if __name__ == "__main__":
    """
    You can modify the code below for debug purpose only.
    Please DONT commit it unless it will cause build break.
    """

    sim = ppsim.Simulator.simple(2, libspu.ProtocolKind.CHEETAH, libspu.FieldType.FM64)
    copts = libspu.CompilerOptions()
    # Tweak compiler options
    copts.disable_div_sqrt_rewrite = True

    x = np.random.randn(3, 4)
    y = np.random.randn(4, 5)
    fn = lambda x, y: jnp.matmul(x, y)

    spu_fn = ppsim.sim_jax(sim, fn, copts=copts)
    z = spu_fn(x, y)

    # print(spu_fn.pphlo)

    print(f"spu out = {z}")
    print(f"cpu out = {fn(x, y)}")
```

```bash
# first run
spu out = [[ 0.6926651  -2.6623726  -1.4583588  -1.3265152   1.6037178 ]
 [ 0.26473236 -2.050335   -0.3489914  -0.4981079  -0.10615158]
 [-0.14936066  1.3455582   0.43166733  1.2879982   0.02372742]]
cpu out = [[ 0.69267344 -2.6623898  -1.4583687  -1.3265169   1.6037189 ]
 [ 0.2647319  -2.0503285  -0.34898427 -0.4981012  -0.10614085]
 [-0.14936455  1.3455673   0.43167225  1.2880112   0.0237307 ]]

# second run
spu out = [[ 0.6926651  -2.6623764  -1.4583588  -1.3265152   1.6037178 ]
 [ 0.26473236 -2.050335   -0.34898758 -0.49810028 -0.1061554 ]
 [-0.14936066  1.345562    0.4316635   1.288002    0.02372742]]
cpu out = [[ 0.69267344 -2.6623898  -1.4583687  -1.3265169   1.6037189 ]
 [ 0.2647319  -2.0503285  -0.34898427 -0.4981012  -0.10614085]
 [-0.14936455  1.3455673   0.43167225  1.2880112   0.0237307 ]]
```

**Comment by zsan2023** [comment]
> Even a simple example of matrix multiplication can have some error after two runs.
> 
> import jax.numpy as jnp
> import numpy as np
> 
> np.random.seed(42)
> 
> import spu.libspu as libspu
> import spu.utils.simulation as ppsim
> 
> if __name__ == "__main__":
>     """
>     You can modify the code below for debug purpose only.
>     Please DONT commit it unless it will cause build break.
>     """
> 
>     sim = ppsim.Simulator.simple(2, libspu.ProtocolKind.CHEETAH, libspu.FieldType.FM64)
>     copts = libspu.CompilerOptions()
>     # Tweak compiler options
>     copts.disable_div_sqrt_rewrite = True
> 
>     x = np.random.randn(3, 4)
>     y = np.random.randn(4, 5)
>     fn = lambda x, y: jnp.matmul(x, y)
> 
>     spu_fn = ppsim.sim_jax(sim, fn, copts=copts)
>     z = spu_fn(x, y)
> 
>     # print(spu_fn.pphlo)
> 
>     print(f"spu out = {z}")
>     print(f"cpu out = {fn(x, y)}")
> # first run
> spu out = [[ 0.6926651  -2.6623726  -1.4583588  -1.3265152   1.6037178 ]
>  [ 0.26473236 -2.050335   -0.3489914  -0.4981079  -0.10615158]
>  [-0.14936066  1.3455582   0.43166733  1.2879982   0.02372742]]
> cpu out = [[ 0.69267344 -2.6623898  -1.4583687  -1.3265169   1.6037189 ]
>  [ 0.2647319  -2.0503285  -0.34898427 -0.4981012  -0.10614085]
>  [-0.14936455  1.3455673   0.43167225  1.2880112   0.0237307 ]]
> 
> # second run
> spu out = [[ 0.6926651  -2.6623764  -1.4583588  -1.3265152   1.6037178 ]
>  [ 0.26473236 -2.050335   -0.34898758 -0.49810028 -0.1061554 ]
>  [-0.14936066  1.345562    0.4316635   1.288002    0.02372742]]
> cpu out = [[ 0.69267344 -2.6623898  -1.4583687  -1.3265169   1.6037189 ]
>  [ 0.2647319  -2.0503285  -0.34898427 -0.4981012  -0.10614085]
>  [-0.14936455  1.3455673   0.43167225  1.2880112   0.0237307 ]]

Ok, thank you for your reply！
